core/threaded/scheduler_GEDF_NP.c

```diff
@@ -0,0 +1,368 @@
+    DEBUG_PRINT("Scheduler: Trying to lock the mutex for level %d.", reaction_level);
+    lf_mutex_lock(&_lf_sched_vector_of_reaction_qs_mutexes[reaction_level]);
+    DEBUG_PRINT("Scheduler: Locked the mutex for level %d.", reaction_level);
+    pqueue_insert(_lf_sched_vector_of_reaction_qs[reaction_level], (void*)reaction);
+    lf_mutex_unlock(&_lf_sched_vector_of_reaction_qs_mutexes[reaction_level]);
+}
+
+/**
+ * @brief Distribute any reaction that is ready to execute to idle worker thread(s).
+ *  
+ * @return Number of reactions that were successfully distributed to worker threads.
+ */ 
+int _lf_sched_distribute_ready_reactions() {
+    pqueue_t* tmp_queue = NULL;
+    // Note: All the threads are idle, which means that they are done inserting
+    // reactions. Therefore, the reaction queues can be accessed without locking
+    // a mutex.
+    for (;_lf_sched_next_reaction_level <= MAX_REACTION_LEVEL; _lf_sched_next_reaction_level++) { 
+        tmp_queue = _lf_sched_vector_of_reaction_qs[_lf_sched_next_reaction_level];
+        size_t reactions_to_execute = pqueue_size(tmp_queue);
+        if (reactions_to_execute) {
+            executing_q = tmp_queue;
+            _lf_sched_next_reaction_level++;
+            return reactions_to_execute;
+        }
+    }
+    
+    return 0;
+}
+
+/**
+ * @brief If there is work to be done, notify workers individually.
+ * 
+ * This assumes that the caller is not holding any thread mutexes.
+ */
+void _lf_sched_notify_workers() {
+    // Note: All threads are idle. Therefore, there is no need to lock the mutex
+    // while accessing the executing queue (which is pointing to one of the
+    // reaction queues).
+    size_t workers_to_be_awaken = MIN(_lf_sched_number_of_idle_workers, pqueue_size(executing_q));
+    DEBUG_PRINT("Scheduler: Notifying %d workers.", workers_to_be_awaken);
+    _lf_sched_number_of_idle_workers -= workers_to_be_awaken;
+    DEBUG_PRINT("Scheduler: New number of idle workers: %u.", _lf_sched_number_of_idle_workers);
+    if (workers_to_be_awaken > 1) {
+        // Notify all the workers except the worker thread that has called this function. 
+        lf_semaphore_release(_lf_sched_semaphore, (workers_to_be_awaken-1));
+    }
+}
+
+/**
+ * @brief Signal all worker threads that it is time to stop.
+ * 
+ */
+void _lf_sched_signal_stop() {
+    _lf_sched_should_stop = true;
+    lf_semaphore_release(_lf_sched_semaphore, (_lf_sched_number_of_workers - 1));
+}
+
+/**
+ * @brief Advance tag or distribute reactions to worker threads.
+ *
+ * Advance tag if there are no reactions on the reaction queue. If
+ * there are such reactions, distribute them to worker threads.
+ * 
+ * This function assumes the caller does not hold the 'mutex' lock.
+ */
+void _lf_sched_try_advance_tag_and_distribute() {
+    // if (pqueue_size(executing_q) != 0) {
+    //     error_print_and_exit("Scheduler: Executing queue is not empty.");
+    // }
```

Comment:
`assert` (or any check) can be expensive, that's why it is commented out entirely. In C++, having a `Release` build type disables asserts, but I'm not sure if it works for the C `assert` macro.
Moreover, `executing_q` is local to the scheduler, so this would not qualify as a precondition. 
We can bite the bullet and add the check.

Reply:
Empirically, it has not appeared to me that `assert` had any overhead outside of debug mode. [This post](https://stackoverflow.com/questions/1571340/what-is-the-assert-function) says it shouldn't, and when I look in `/usr/include/assert.h`, I see handling of `NDEBUG` that looks right, although I don't understand everything that happens.

Reply:
Okay, that's promising. Then we can safely go with an assert here.